### PR TITLE
releng - docker - sign docker images

### PR DESCRIPTION
**Pull request by thisisshi**
Uses the cosign experimental keyless signing, used this as a reference: https://github.com/chrisns/cosign-keyless-demo

### CODE REVIEW

File: .github/composites/docker-build-push/action.yaml [hunk 1]
```diff
@@ -144,3 +146,24 @@ runs:
...
+      if: ${{ inputs.push == true }}
+      shell: bash
+      env:
+        COSIGN_EXPERIMENTAL: "true"
+        TAGS: ${{ steps.meta.outputs.tags }}
+      run: |
+        tags=(${TAGS//,/ })
+        for i in "${tags[@]}"
+        do
+          docker pull $i
```
Comment (kapilt) [inline .github/composites/docker-build-push/action.yaml:166]: is pull needed? sign is typically done against a local artifact, pulling an "unknown" artifact to sign it, seems problematic.

File: .github/composites/docker-build-push/action.yaml [hunk 2]
```diff
@@ -144,3 +146,24 @@ runs:
...
+          cosign sign $i
+          cosign verify $i
```
Comment (kapilt) [inline .github/composites/docker-build-push/action.yaml:168]: why re verify, it was just signed?

File: .github/composites/docker-build-push/action.yaml (hunk 1 above)
Comment (thisisshi) [reply .github/composites/docker-build-push/action.yaml:166]: this is a limitation on the buildx side, we're unable to export images to docker from buildx if it's a multiplatform build: https://github.com/docker/buildx/issues/59 I'm open to ideas on how to get around this, one comment said to push to a local registry which might be an option

File: .github/composites/docker-build-push/action.yaml (hunk 2 above)
Comment (thisisshi) [reply .github/composites/docker-build-push/action.yaml:168]: stray debug line

File: .github/composites/docker-build-push/action.yaml (hunk 1 above)
Comment (kapilt) [reply .github/composites/docker-build-push/action.yaml:166]: Worth exploring if we can use co-sign directly against registry, else pull will launder multi arch to single arch, and we want to sign the multi arch manifest. Also worth an issue upstream if its not directly supported per exploration.

Comment (thisisshi) [reply .github/composites/docker-build-push/action.yaml]: Yup, was able to do a remote signing, example:

```
COSIGN_EXPERIMENTAL=true cosign verify sonnyshi/c7n:docker-buildx4

Verification for index.docker.io/sonnyshi/c7n:docker-buildx4 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - Any certificates were verified against the Fulcio roots.

[{"critical":{"identity":{"docker-reference":"index.docker.io/sonnyshi/c7n"},"image":{"docker-manifest-digest":"sha256:31693ffb921b3f32fc55d3ba51061aa47e74207374b256ad86cbe5e2c3e75b3e"},"type":"cosign container image signature"},"optional":{"Bundle":{"SignedEntryTimestamp":"MEUCICC2d1Zh7wa19m7m7AyR+EM2vL9AJYmi3eI30i2HIHaqAiEA9tRlVT1aTyWTrLM/swN3bbX1MvMlXwXskdFA1Z8kiQY=","Payload":{"body":"eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiJkNTgxODgyZDU5NjEyOTBjN2ZkOGMwYzZmZjliYzk5NzUxNzQ3ZDgxZjUyZmFlYjY4Mjc5YjQ1Y2EyN2YzNmViIn19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FUUNJRjVNWk40b1VKRXhPbFd0WUI4c2pJZno4Y3lSTmpYcis3M1ZvVHV6WnRjQUFpQTZCU0dacWdFUitpc1VETFUwYi9PaG5uaUFKRnRueStoZE5uaDdIVldJQ0E9PSIsInB1YmxpY0tleSI6eyJjb250ZW50IjoiTFMwdExTMUNSVWRKVGlCRFJWSlVTVVpKUTBGVVJTMHRMUzB0Q2sxSlNVTnVha05EUVdsVFowRjNTVUpCWjBsVlptUmxXRXBxTm00MmEyOTRUVlJ5U21seVJGWTVNVU5MU2xKamQwTm5XVWxMYjFwSmVtb3dSVUYzVFhjS1RucEZWazFDVFVkQk1WVkZRMmhOVFdNeWJHNWpNMUoyWTIxVmRWcEhWakpOVWpSM1NFRlpSRlpSVVVSRmVGWjZZVmRrZW1SSE9YbGFVekZ3WW01U2JBcGpiVEZzV2tkc2FHUkhWWGRJYUdOT1RXcEpkMDlVUVhsTmFrVjZUbFJOZDFkb1kwNU5ha2wzVDFSQmVVMXFSVEJPVkUxM1YycEJRVTFHYTNkRmQxbElDa3R2V2tsNmFqQkRRVkZaU1V0dldrbDZhakJFUVZGalJGRm5RVVZ3V25WRVJFdFBWblZNUTBWTVFtSmhVbVJQTTJkblltWjNWMjkxYUd3ek1WRmhNeThLTURVM1RtTjRkMXBDYTJkNmVFVlJMM1JNZW5wVU5XSmpZbTFZTjNGRVZsVm5OVGhFUTJKVVEzZzRTRFE1Ym14cVR6WlBRMEZWVFhkblowVXZUVUUwUndwQk1WVmtSSGRGUWk5M1VVVkJkMGxJWjBSQlZFSm5UbFpJVTFWRlJFUkJTMEpuWjNKQ1owVkdRbEZqUkVGNlFXUkNaMDVXU0ZFMFJVWm5VVlZKSzIxM0NuaEJhRlpVYUdSVllXVXpXbVV3UmxsUmNtWTVablJqZDBoM1dVUldVakJxUWtKbmQwWnZRVlV6T1ZCd2VqRlphMFZhWWpWeFRtcHdTMFpYYVhocE5Ga0tXa1E0ZDBoUldVUldVakJTUVZGSUwwSkNUWGRGV1VWUVlVZHNRV015T1hWaWJteDZZVWRyZFZreU9YUk5RM2RIUTJselIwRlJVVUpuTnpoM1FWRkZSUXBJYldnd1pFaENlazlwT0haYU1td3dZVWhXYVV4dFRuWmlVemx6WWpKa2NHSnBPWFpaV0ZZd1lVUkRRbWxuV1V0TGQxbENRa0ZJVjJWUlNVVkJaMUk0Q2tKSWIwRmxRVUl5UVVGb1oydDJRVzlWZGpsdlVtUklVbUY1WlVWdVJWWnVSMHQzVjFCalRUUXdiVE50ZGtOSlIwNXRPWGxCUVVGQ1ozZEJhRmgyTUVFS1FVRlJSRUZGWTNkU1VVbG5UMlEzYWpkblFtUktWekpGTjFabFNVcE5TalVyTUhRclZXRkxlbVUwWTJwSmRURkRNV3hKYVZGSGQwTkpVVU5XUm1NMll3b3hTM1JDVDNNNU56UnFkMlpvUlROV2MwUnRRV2Q2U1RodmFYRklWa05yTDI1M1pWZHpha0ZMUW1kbmNXaHJhazlRVVZGRVFYZE9iMEZFUW14QmFrRm9DbmRxTWxWT2RXRktjazVTT0hoSVJXdFNkVTl5ZVZkWmJGVkhRM2xxT0dSdE5uUjNjMDFGYkZnNVIwaElaaXM0UTNFdmRtVk1RMUpvVjJsMmIyOHlkME1LVFZGRWJXOTJUbm92WmxOalRETmpaRmcxYXpKeloyOHpjR3B2VDNrNUswWXhaR1JZTlVsNGRIQmpja0ZRUzBkelNVdGtkVkZ6WlhFdlpVSm5kVzlYYWdvMldsazlDaTB0TFMwdFJVNUVJRU5GVWxSSlJrbERRVlJGTFMwdExTMEsifX19fQ==","integratedTime":1662154531,"logIndex":3369119,"logID":"c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d"}},"Issuer":"https://github.com/login/oauth","Subject":"hi@sonnyshi.com"}}]
```

From my fork: https://github.com/thisisshi/cloud-custodian/runs/8164256449?check_suite_focus=true

File: .github/composites/docker-build-push/action.yaml [hunk 3]
```diff
@@ -144,3 +146,22 @@ runs:
...
+      uses: sigstore/cosign-installer@main
+      with:
+        cosign-release: 'v1.11.0'
+
+    - name: Sign
+      if: "${{ inputs.push == 'true' }}"
+      shell: bash
+      env:
+        COSIGN_EXPERIMENTAL: "true"
+        TAGS: ${{ steps.meta.outputs.tags }}
```
Comment (kapilt) [inline .github/composites/docker-build-push/action.yaml:161]: what's the echo on this value look like?

Comment (thisisshi) [reply .github/composites/docker-build-push/action.yaml:161]: this should look like `latest,0.9.19.0`

**Comment by kapilt** [comment]
looks like 2.5.1 of the action isn't found, missing a `v` prefix I presume, I would probably switch to @v2, we'll separately be looking at pinning across our actions to git shas similar to what was done for tfparse with ratchet. https://github.com/cloud-custodian/tfparse/pull/20

**Review by kapilt — APPROVED** [review]
lgtm

File: .github/composites/docker-build-push/action.yaml [hunk 4]
```diff
@@ -63,7 +65,7 @@ runs:
         images: |
           ${{ inputs.repository }}/${{ inputs.name }}
         tags: |
-          type=semver,pattern={raw}
+          type=ref,event=tag
```
Comment (thisisshi) [inline .github/composites/docker-build-push/action.yaml:68]: our versioning schema isn't semver compatible

**Pull request merged**